File: internal/handler/vectora.go
package handler

import (
	"context"

	"connectrpc.com/connect"

	omegav1 "github.com/benebsworth/omega/gen/go/omega/v1"
	omegav1connect "github.com/benebsworth/omega/gen/go/omega/v1/omegav1connect"
	"github.com/benebsworth/omega/internal/db"
)

// Ensure interface satisfaction at compile time.
var _ omegav1connect.VectoraServiceHandler = (*VectoraHandler)(nil)

// VectoraHandler implements VectoraService.
type VectoraHandler struct {
	vdb *db.VectoraDB
}

// NewVectora creates a VectoraHandler backed by the given VectoraDB.
func NewVectora(vdb *db.VectoraDB) *VectoraHandler {
	return &VectoraHandler{vdb: vdb}
}

// ── GetPortfolio ──────────────────────────────────────────────────────────────

func (h *VectoraHandler) GetPortfolio(
	ctx context.Context,
	req *connect.Request[omegav1.GetPortfolioRequest],
) (*connect.Response[omegav1.GetPortfolioResponse], error) {
	p, err := h.vdb.GetPortfolio()
	if err != nil {
		return nil, connect.NewError(connect.CodeInternal, err)
	}

	var alloc []*omegav1.VectoraAllocationSlice
	for _, a := range p.Allocation {
		alloc = append(alloc, &omegav1.VectoraAllocationSlice{
			Name:  a.Name,
			Value: a.Value,
			Color: a.Color,
		})
	}

	return connect.NewResponse(&omegav1.GetPortfolioResponse{
		PortfolioValue: p.PortfolioValue,
		UnrealisedPnl:  p.UnrealisedPnL,
		RealisedPnl:    p.RealisedPnL,
		TotalPnl:       p.TotalPnL,
		TotalReturn:    p.TotalReturn,
		AnnReturn:      p.AnnReturn,
		WinRate:        p.WinRate,
		ProfitFactor:   p.ProfitFactor,
		Sharpe:         p.Sharpe,
		AnnVol:         p.AnnVol,
		Allocation:     alloc,
	}), nil
}

// ── GetPositions ──────────────────────────────────────────────────────────────

func (h *VectoraHandler) GetPositions(
	ctx context.Context,
	req *connect.Request[omegav1.GetPositionsRequest],
) (*connect.Response[omegav1.GetPositionsResponse], error) {
	positions, err := h.vdb.GetPositions()
	if err != nil {
		return nil, connect.NewError(connect.CodeInternal, err)
	}

	var protoPos []*omegav1.VectoraPosition
	for _, p := range positions {
		protoPos = append(protoPos, &omegav1.VectoraPosition{
			Sym:      p.Sym,
			Side:     p.Side,
			Size:     p.Size,
			Entry:    p.Entry,
			Mark:     p.Mark,
			Upnl:     p.Upnl,
			Pct:      p.Pct,
			Notional: p.Notional,
			Leverage: p.Leverage,
			Var95:    p.Var95,
		})
	}

	return connect.NewResponse(&omegav1.GetPositionsResponse{Positions: protoPos}), nil
}

// ── GetPnL ────────────────────────────────────────────────────────────────────

func (h *VectoraHandler) GetPnL(
	ctx context.Context,
	req *connect.Request[omegav1.GetPnLRequest],
) (*connect.Response[omegav1.GetPnLResponse], error) {
	p, err := h.vdb.GetPnL()
	if err != nil {
		return nil, connect.NewError(connect.CodeInternal, err)
	}

	return connect.NewResponse(&omegav1.GetPnLResponse{
		UnrealisedPnl: p.UnrealisedPnL,
		RealisedPnl:   p.RealisedPnL,
		TotalPnl:      p.TotalPnL,
		TotalReturn:   p.TotalReturn,
		AnnReturn:     p.AnnReturn,
		WinRate:       p.WinRate,
		ProfitFactor:  p.ProfitFactor,
		Sharpe:        p.Sharpe,
		AnnVol:        p.AnnVol,
		MaxDd:         p.MaxDD,
		Var95:         p.Var95,
		Cvar95:        p.CVaR95,
		Sortino:       p.Sortino,
		Calmar:        p.Calmar,
	}), nil
}

// ── GetSignals ────────────────────────────────────────────────────────────────

func (h *VectoraHandler) GetSignals(
	ctx context.Context,
	req *connect.Request[omegav1.GetSignalsRequest],
) (*connect.Response[omegav1.GetSignalsResponse], error) {
	signals, err := h.vdb.GetSignals()
	if err != nil {
		return nil, connect.NewError(connect.CodeInternal, err)
	}

	var protoSigs []*omegav1.VectoraSignal
	var totalConviction, totalWeight float64
	for _, s := range signals {
		protoSigs = append(protoSigs, &omegav1.VectoraSignal{
			Name:         s.Name,
			AvgIc:        s.AvgIC,
			Weight:       s.Weight,
			HalfLife:     int32(s.HalfLife), //nolint:gosec
			Color:        s.Color,
			Conviction:   s.Conviction,
			BrierScore:   s.BrierScore,
			CurrentValue: s.CurrentValue,
			Trend:        s.Trend,
		})
		totalConviction += s.Conviction * s.Weight
		totalWeight += s.Weight
	}

	compositeScore := 0.0
	if totalWeight > 0 {
		compositeScore = totalConviction / totalWeight
	}
	compositeDir := "NEUTRAL"
	if compositeScore > 0.5 {
		compositeDir = "LONG"
	} else if compositeScore < 0.4 {
		compositeDir = "SHORT"
	}

	// OOS Sharpe from backtest table
	bt, _ := h.vdb.GetBacktestStats()
	oosSharpe := 0.0
	if bt != nil {
		oosSharpe = bt.SharpeOOS
	}

	return connect.NewResponse(&omegav1.GetSignalsResponse{
		Signals:            protoSigs,
		CompositeScore:     compositeScore,
		CompositeDirection: compositeDir,
		OosSharpe:          oosSharpe,
	}), nil
}

// ── GetSignalHistory ──────────────────────────────────────────────────────────

func (h *VectoraHandler) GetSignalHistory(
	ctx context.Context,
	req *connect.Request[omegav1.GetSignalHistoryRequest],
) (*connect.Response[omegav1.GetSignalHistoryResponse], error) {
	points, err := h.vdb.GetSignalHistory(req.Msg.SignalName, int(req.Msg.Limit))
	if err != nil {
		return nil, connect.NewError(connect.CodeInternal, err)
	}

	var protoPoints []*omegav1.VectoraSignalICPoint
	for _, p := range points {
		protoPoints = append(protoPoints, &omegav1.VectoraSignalICPoint{
			T:  int32(p.T), //nolint:gosec
			Ic: p.IC,
		})
	}

	return connect.NewResponse(&omegav1.GetSignalHistoryResponse{Points: protoPoints}), nil
}

// ── GetTrades ─────────────────────────────────────────────────────────────────

func (h *VectoraHandler) GetTrades(
	ctx context.Context,
	req *connect.Request[omegav1.GetTradesRequest],
) (*connect.Response[omegav1.GetTradesResponse], error) {
	trades, err := h.vdb.GetTrades(req.Msg.SymFilter, req.Msg.SideFilter, int(req.Msg.Limit))
	if err != nil {
		return nil, connect.NewError(connect.CodeInternal, err)
	}

	var protoTrades []*omegav1.VectoraTrade
	for _, t := range trades {
		protoTrades = append(protoTrades, &omegav1.VectoraTrade{
			Ts:         t.Ts,
			Sym:        t.Sym,
			Side:       t.Side,
			Size:       t.Size,
			Entry:      t.Entry,
			ExitPrice:  t.ExitPrice,
			Pnl:        t.Pnl,
			Slippage:   t.Slippage,
			Duration:   t.Duration,
		})
	}

	return connect.NewResponse(&omegav1.GetTradesResponse{Trades: protoTrades}), nil
}

// ── GetBacktestResults ────────────────────────────────────────────────────────

func (h *VectoraHandler) GetBacktestResults(
	ctx context.Context,
	req *connect.Request[omegav1.GetBacktestResultsRequest],
) (*connect.Response[omegav1.GetBacktestResultsResponse], error) {
	s, err := h.vdb.GetBacktestStats()
	if err != nil {
		return nil, connect.NewError(connect.CodeInternal, err)
	}

	return connect.NewResponse(&omegav1.GetBacktestResultsResponse{
		Stats: &omegav1.VectoraBacktestStats{
			SharpeAnn:     s.SharpeAnn,
			SortinoAnn:    s.SortinoAnn,
			MaxDdPct:      s.MaxDDPct,
			Calmar:        s.Calmar,
			SharpeIs:      s.SharpeIS,
			SharpeOos:     s.SharpeOOS,
			Var:           s.VaR,
			Cvar:          s.CVaR,
			MeanR:         s.MeanR,
			StdR:          s.StdR,
			AnnReturn:     s.AnnReturn,
			TotalReturn:   s.TotalReturn,
			PortfolioValue: s.PortfolioVal,
			MaxDdDuration: int32(s.MaxDDDuration), //nolint:gosec
			WinRate:       s.WinRate,
			ProfitFactor:  s.ProfitFactor,
			TrainEnd:      int32(s.TrainEnd), //nolint:gosec
		},
	}), nil
}

// ── GetEquityCurve ────────────────────────────────────────────────────────────

func (h *VectoraHandler) GetEquityCurve(
	ctx context.Context,
	req *connect.Request[omegav1.GetEquityCurveRequest],
) (*connect.Response[omegav1.GetEquityCurveResponse], error) {
	points, trainEnd, err := h.vdb.GetEquityCurve(int(req.Msg.Limit))
	if err != nil {
		return nil, connect.NewError(connect.CodeInternal, err)
	}

	var protoPoints []*omegav1.VectoraEquityPoint
	for _, p := range points {
		protoPoints = append(protoPoints, &omegav1.VectoraEquityPoint{
			Date:  p.Date,
			I:     int32(p.I), //nolint:gosec
			Omega: p.Omega,
			Btc:   p.Btc,
			Dd:    p.DD,
		})
	}

	return connect.NewResponse(&omegav1.GetEquityCurveResponse{
		Points:   protoPoints,
		TrainEnd: int32(trainEnd), //nolint:gosec
	}), nil
}

// ── GetRiskMetrics ────────────────────────────────────────────────────────────

func (h *VectoraHandler) GetRiskMetrics(
	ctx context.Context,
	req *connect.Request[omegav1.GetRiskMetricsRequest],
) (*connect.Response[omegav1.GetRiskMetricsResponse], error) {
	m, err := h.vdb.GetRiskMetrics()
	if err != nil {
		return nil, connect.NewError(connect.CodeInternal, err)
	}

	var ablation []*omegav1.VectoraAblationEntry
	for _, a := range m.Ablation {
		ablation = append(ablation, &omegav1.VectoraAblationEntry{
			Name:    a.Name,
			DSharpe: a.DSharpe,
			Sig:     a.Sig,
		})
	}

	var regimes []*omegav1.VectoraRegime
	for _, r := range m.Regimes {
		regimes = append(regimes, &omegav1.VectoraRegime{
			Name:   r.Name,
			Sharpe: r.Sharpe,
			Ret:    r.Ret,
			Trades: int32(r.Trades), //nolint:gosec
			Pct:    r.Pct,
		})
	}

	var crashes []*omegav1.VectoraCrashScenario
	for _, c := range m.Crashes {
		sc := &omegav1.VectoraCrashScenario{
			Name:     c.Name,
			Sym:      c.Sym,
			Dd:       c.DD,
			Sl:       int32(c.SL), //nolint:gosec
			Pnl:      c.Pnl,
			Pass:     c.Pass,
			HasRecov: c.Recov != nil,
		}
		if c.Recov != nil {
			sc.Recov = *c.Recov
		}
		crashes = append(crashes, sc)
	}

	var funding []*omegav1.VectoraFundingPoint
	for _, f := range m.FundingData {
		funding = append(funding, &omegav1.VectoraFundingPoint{
			T:       int32(f.T), //nolint:gosec
			Funding: f.Funding,
			Oi:      f.OI,
		})
	}

	var adv []*omegav1.VectoraAdvPoint
	for _, a := range m.AdvSeries {
		adv = append(adv, &omegav1.VectoraAdvPoint{
			T:    int32(a.T), //nolint:gosec
			Flag: a.Flag,
			Fp:   a.FP,
		})
	}

	var tpe []*omegav1.VectoraTpePoint
	for _, t := range m.TpeSeries {
		tpe = append(tpe, &omegav1.VectoraTpePoint{
			T:     int32(t.T), //nolint:gosec
			Score: t.Score,
			Best:  t.Best,
		})
	}

	return connect.NewResponse(&omegav1.GetRiskMetricsResponse{
		Ablation:         ablation,
		Regimes:          regimes,
		CurrentRegimeIdx: int32(m.CurrentRegimeIdx), //nolint:gosec
		Crashes:          crashes,
		FundingData:      funding,
		AdvSeries:        adv,
		TpeSeries:        tpe,
		LatestFunding:    m.LatestFunding,
		LatestOi:         m.LatestOI,
	}), nil
}

File: dashboard/src/api/vectora.ts
/**
 * Vectora API client — wraps Connect-ES calls to VectoraService.
 * Falls back to mock data when the API is unreachable.
 */
import { createClient } from "@connectrpc/connect";
import { createConnectTransport } from "@connectrpc/connect-web";
import { VectoraService } from "../gen/omega/v1/vectora_service_pb";
import type {
  AllocationSlice,
  BacktestStats,
  PerfDataPoint,
  Position,
  Signal,
  Trade,
  AblationEntry,
  Regime,
  CrashScenario,
  FundingPoint,
  AdvPoint,
  TpePoint,
} from "../mocks/vectora";
import * as mock from "../mocks/vectora";

// ---------------------------------------------------------------------------
// Client setup
// ---------------------------------------------------------------------------

const baseUrl = import.meta.env.VITE_API_URL ?? "";

const transport = createConnectTransport({ baseUrl });
const vectoraClient = createClient(VectoraService, transport);

// ---------------------------------------------------------------------------
// API availability state — set once on first failure, cleared on success
// ---------------------------------------------------------------------------

let _apiAvailable: boolean | null = null; // null = untested

export function isApiAvailable(): boolean {
  return _apiAvailable === true;
}

async function callApi<T>(
  fn: () => Promise<T>
): Promise<{ data: T; fromMock: false } | { data: null; fromMock: true }> {
  try {
    const data = await fn();
    _apiAvailable = true;
    return { data, fromMock: false };
  } catch {
    _apiAvailable = false;
    return { data: null, fromMock: true };
  }
}

// ---------------------------------------------------------------------------
// Portfolio
// ---------------------------------------------------------------------------

export interface PortfolioData {
  portfolioValue: number;
  unrealisedPnl: number;
  realisedPnl: number;
  totalPnl: number;
  totalReturn: number;
  annReturn: number;
  winRate: number;
  profitFactor: number;
  sharpe: number;
  annVol: number;
  allocation: AllocationSlice[];
  fromMock: boolean;
}

export async function fetchPortfolio(): Promise<PortfolioData> {
  const result = await callApi(() => vectoraClient.getPortfolio({}));

  if (result.fromMock) {
    const positions = mock.positions;
    const totalNotional = positions.reduce((s, p) => s + p.notional, 0);
    const cashValue = mock.allocation.find((a) => a.name === "USDT Cash")?.value ?? 0;
    const portfolioValue = totalNotional + cashValue;
    return {
      portfolioValue,
      unrealisedPnl: positions.reduce((s, p) => s + p.upnl, 0),
      realisedPnl: mock.backtestStats.totalReturn * portfolioValue * 0.1,
      totalPnl: positions.reduce((s, p) => s + p.upnl, 0),
      totalReturn: mock.backtestStats.totalReturn,
      annReturn: mock.backtestStats.annReturn,
      winRate: mock.backtestStats.winRate,
      profitFactor: mock.backtestStats.profitFactor,
      sharpe: mock.backtestStats.sharpeAnn,
      annVol: mock.backtestStats.stdR * Math.sqrt(252),
      allocation: mock.allocation,
      fromMock: true,
    };
  }

  const d = result.data;
  return {
    portfolioValue: d.portfolioValue,
    unrealisedPnl: d.unrealisedPnl,
    realisedPnl: d.realisedPnl,
    totalPnl: d.totalPnl,
    totalReturn: d.totalReturn,
    annReturn: d.annReturn,
    winRate: d.winRate,
    profitFactor: d.profitFactor,
    sharpe: d.sharpe,
    annVol: d.annVol,
    allocation: d.allocation.map((a) => ({
      name: a.name,
      value: a.value,
      color: a.color,
    })),
    fromMock: false,
  };
}

// ---------------------------------------------------------------------------
// Positions
// ---------------------------------------------------------------------------

export async function fetchPositions(): Promise<{ positions: Position[]; fromMock: boolean }> {
  const result = await callApi(() => vectoraClient.getPositions({}));

  if (result.fromMock) {
    return { positions: mock.positions, fromMock: true };
  }

  return {
    positions: result.data.positions.map((p) => ({
      sym: p.sym,
      side: p.side as "LONG" | "SHORT",
      size: p.size,
      entry: p.entry,
      mark: p.mark,
      upnl: p.upnl,
      pct: p.pct,
      notional: p.notional,
      leverage: p.leverage,
      var95: p.var95,
    })),
    fromMock: false,
  };
}

// ---------------------------------------------------------------------------
// PnL
// ---------------------------------------------------------------------------

export interface PnLData {
  unrealisedPnl: number;
  realisedPnl: number;
  totalPnl: number;
  totalReturn: number;
  annReturn: number;
  winRate: number;
  profitFactor: number;
  sharpe: number;
  annVol: number;
  maxDD: number;
  var95: number;
  cvar95: number;
  sortino: number;
  calmar: number;
  fromMock: boolean;
}

export async function fetchPnL(): Promise<PnLData> {
  const result = await callApi(() => vectoraClient.getPnL({}));

  if (result.fromMock) {
    const s = mock.backtestStats;
    const totalNotional = mock.positions.reduce((sum, p) => sum + p.notional, 0);
    return {
      unrealisedPnl: mock.positions.reduce((sum, p) => sum + p.upnl, 0),
      realisedPnl: s.totalReturn * totalNotional * 0.1,
      totalPnl: mock.positions.reduce((sum, p) => sum + p.upnl, 0),
      totalReturn: s.totalReturn,
      annReturn: s.annReturn,
      winRate: s.winRate,
      profitFactor: s.profitFactor,
      sharpe: s.sharpeAnn,
      annVol: s.stdR * Math.sqrt(252),
      maxDD: s.maxDDpct,
      var95: s.VaR,
      cvar95: s.CVaR,
      sortino: s.sortinoAnn,
      calmar: s.calmar,
      fromMock: true,
    };
  }

  const d = result.data;
  return {
    unrealisedPnl: d.unrealisedPnl,
    realisedPnl: d.realisedPnl,
    totalPnl: d.totalPnl,
    totalReturn: d.totalReturn,
    annReturn: d.annReturn,
    winRate: d.winRate,
    profitFactor: d.profitFactor,
    sharpe: d.sharpe,
    annVol: d.annVol,
    maxDD: d.maxDd,
    var95: d.var95,
    cvar95: d.cvar95,
    sortino: d.sortino,
    calmar: d.calmar,
    fromMock: false,
  };
}

// ---------------------------------------------------------------------------
// Signals
// ---------------------------------------------------------------------------

export interface SignalsData {
  signals: Signal[];
  compositeScore: number;
  compositeDirection: string;
  oosSharpe: number;
  fromMock: boolean;
}

export async function fetchSignals(): Promise<SignalsData> {
  const result = await callApi(() => vectoraClient.getSignals({}));

  if (result.fromMock) {
    const weighted = mock.signals.reduce((s, sig) => s + sig.conviction * sig.weight, 0);
    const totalW = mock.signals.reduce((s, sig) => s + sig.weight, 0);
    const composite = totalW > 0 ? weighted / totalW : 0;
    return {
      signals: mock.signals,
      compositeScore: composite,
      compositeDirection: composite > 0.5 ? "LONG" : composite < 0.4 ? "SHORT" : "NEUTRAL",
      oosSharpe: mock.backtestStats.sharpeOOS,
      fromMock: true,
    };
  }

  const d = result.data;
  return {
    signals: d.signals.map((s) => ({
      name: s.name,
      avgIC: s.avgIc,
      weight: s.weight,
      halfLife: s.halfLife,
      color: s.color,
      conviction: s.conviction,
      brierScore: s.brierScore,
      currentValue: s.currentValue,
      trend: s.trend as "up" | "down" | "flat",
    })),
    compositeScore: d.compositeScore,
    compositeDirection: d.compositeDirection,
    oosSharpe: d.oosSharpe,
    fromMock: false,
  };
}

export async function fetchSignalHistory(
  signalName: string,
  limit = 60
): Promise<{ points: { t: number; ic: number }[]; fromMock: boolean }> {
  const result = await callApi(() => vectoraClient.getSignalHistory({ signalName, limit }));

  if (result.fromMock) {
    // Generate deterministic IC history from mock signal
    const sig = mock.signals.find((s) => s.name === signalName);
    const baseIC = sig?.avgIC ?? 0.05;
    const points = Array.from({ length: limit }, (_, i) => ({
      t: i + 1,
      ic: +(baseIC + 0.03 * Math.sin(i / 5) + 0.015 * (Math.random() - 0.5)).toFixed(4),
    }));
    return { points, fromMock: true };
  }

  return {
    points: result.data.points.map((p) => ({ t: p.t, ic: p.ic })),
    fromMock: false,
  };
}

// ---------------------------------------------------------------------------
// Trades
// ---------------------------------------------------------------------------

export async function fetchTrades(
  symFilter = "",
  sideFilter = "",
  limit = 100
): Promise<{ trades: Trade[]; fromMock: boolean }> {
  const result = await callApi(() => vectoraClient.getTrades({ symFilter, sideFilter, limit }));

  if (result.fromMock) {
    let trades = mock.trades;
    if (symFilter) trades = trades.filter((t) => t.sym === symFilter);
    if (sideFilter) trades = trades.filter((t) => t.side === sideFilter);
    return { trades, fromMock: true };
  }

  return {
    trades: result.data.trades.map((t) => ({
      ts: t.ts,
      sym: t.sym,
      side: t.side as "LONG" | "SHORT",
      size: t.size,
      entry: t.entry,
      exit: t.exitPrice,
      pnl: t.pnl,
      slippage: t.slippage,
      duration: t.duration,
    })),
    fromMock: false,
  };
}

// ---------------------------------------------------------------------------
// Backtest
// ---------------------------------------------------------------------------

export async function fetchBacktestResults(): Promise<{
  stats: BacktestStats;
  trainEnd: number;
  fromMock: boolean;
}> {
  const result = await callApi(() => vectoraClient.getBacktestResults({}));

  if (result.fromMock) {
    return { stats: mock.backtestStats, trainEnd: mock.TRAIN_END, fromMock: true };
  }

  const s = result.data.stats;
  if (!s) return { stats: mock.backtestStats, trainEnd: mock.TRAIN_END, fromMock: true };

  return {
    stats: {
      sharpeAnn: s.sharpeAnn,
      sortinoAnn: s.sortinoAnn,
      maxDDpct: s.maxDdPct,
      calmar: s.calmar,
      sharpeIS: s.sharpeIs,
      sharpeOOS: s.sharpeOos,
      VaR: s.var,
      CVaR: s.cvar,
      meanR: s.meanR,
      stdR: s.stdR,
      annReturn: s.annReturn,
      totalReturn: s.totalReturn,
      portfolioValue: s.portfolioValue,
      maxDDDuration: s.maxDdDuration,
      winRate: s.winRate,
      profitFactor: s.profitFactor,
    },
    trainEnd: s.trainEnd,
    fromMock: false,
  };
}

// ---------------------------------------------------------------------------
// Equity curve
// ---------------------------------------------------------------------------

export async function fetchEquityCurve(
  limit = 0
): Promise<{ points: PerfDataPoint[]; trainEnd: number; fromMock: boolean }> {
  const result = await callApi(() => vectoraClient.getEquityCurve({ limit }));

  if (result.fromMock) {
    return { points: mock.perfData, trainEnd: mock.TRAIN_END, fromMock: true };
  }

  return {
    points: result.data.points.map((p) => ({
      date: p.date,
      i: p.i,
      omega: p.omega,
      btc: p.btc,
      dd: p.dd,
    })),
    trainEnd: result.data.trainEnd,
    fromMock: false,
  };
}

// ---------------------------------------------------------------------------
// Risk metrics
// ---------------------------------------------------------------------------

export interface RiskMetricsData {
  ablation: AblationEntry[];
  regimes: Regime[];
  currentRegimeIdx: number;
  crashes: CrashScenario[];
  fundingData: FundingPoint[];
  advSeries: AdvPoint[];
  tpeSeries: TpePoint[];
  latestFunding: number;
  latestOI: number;
  fromMock: boolean;
}

export async function fetchRiskMetrics(): Promise<RiskMetricsData> {
  const result = await callApi(() => vectoraClient.getRiskMetrics({}));

  if (result.fromMock) {
    return {
      ablation: mock.ablation,
      regimes: mock.regimes,
      currentRegimeIdx: mock.currentRegimeIdx,
      crashes: mock.crashes,
      fundingData: mock.fundingData,
      advSeries: mock.advSeries,
      tpeSeries: mock.tpeSeries,
      latestFunding: mock.latestFunding,
      latestOI: mock.latestOI,
      fromMock: true,
    };
  }

  const d = result.data;
  return {
    ablation: d.ablation.map((a) => ({ name: a.name, dSharpe: a.dSharpe, sig: a.sig })),
    regimes: d.regimes.map((r) => ({
      name: r.name,
      sharpe: r.sharpe,
      ret: r.ret,
      trades: r.trades,
      pct: r.pct,
    })),
    currentRegimeIdx: d.currentRegimeIdx,
    crashes: d.crashes.map((c) => ({
      name: c.name,
      sym: c.sym,
      dd: c.dd,
      recov: c.hasRecov ? c.recov : null,
      sl: c.sl,
      pnl: c.pnl,
      pass: c.pass,
    })),
    fundingData: d.fundingData.map((f) => ({ t: f.t, funding: f.funding, oi: f.oi })),
    advSeries: d.advSeries.map((a) => ({ t: a.t, flag: a.flag, fp: a.fp })),
    tpeSeries: d.tpeSeries.map((t) => ({ t: t.t, score: t.score, best: t.best })),
    latestFunding: d.latestFunding,
    latestOI: d.latestOi,
    fromMock: false,
  };
}
